README.md
# SPARE
### Static Page AJAX to Replace Elements — a lightweight client-side library

- *Release 1 was on March 24, 2015, under the terms of the Apache 2.0 license.*
- *Release 2 was on October 16, 2019 — same terms.  Added simulateNavigation.*
- *Release 3 was on June 12, 2021 — same terms.  Improved simulateNavigation.*
- *Release 4 was on August 3, 2022, with a new BSD-like license.  Improved onPopStateRestore.*

SPARE is a small client-side AJAX framework which requires no server-side support.  In fact, the case it’s optimized for is when the server provides only plain static HTML content.  It’s also super easy to use: you only need to call one method.

How it’s designed to work was originally inspired by ASP&#46;Net’s `UpdatePanel` control, but as implemented, it’s more similar to jQuery’s `load()` method, at lower cost (and as a tradeoff, less support for downlevel browsers than 1.x versions of jQuery).

To use it, you give it the ID of any element on your page (which we will call the target), the URL of a new page, and an ID on that page.  It replaces the contents of the local document’s target element with the content of the named element on the new page.  SPARE is most seamlessly used when you have a site design where many pages share common markup for headers, navigation, and so on, and the real differences are in a bounded content area.  SPARE lets you load new content into that area without refreshing the rest of the page.  In this use case, often the two IDs will be the same.

In that scenario, you can ask SPARE to fully simulate navigation as if the pages were being loaded normally instead of partially.  This mode is used by invoking a different method.  In that mode you must attach a `popstate` event handler to support use of the Back button.

You can just as easily select content from pages not resembling the calling page.  You can optionally send POST data as well (though not with simulated navigation at this time).  And there’s an option to invoke callback functions on success or failure.  (We will support a promise-based interface in a future version.)  None of these is required for basic usage.

And if the URL you give returns a page fragment, so you don’t have to select an element within it, that’s even simpler.  That mode works for text content that isn’t even HTML (but don’t try it with binary content, such as an image URL, or it will just look like a mess).

The Javascript API consists of an object named **`SPARE`** with five public methods.  Note that you do not use a `new` operator to instantiate SPARE; it’s a singleton static object.

--------

### replaceContent method

The main method for general use is **`SPARE.replaceContent`**, which takes the following arguments, all of string type unless stated otherwise:

> **`target`** (required): either the DOM ID of the target element in your document, or a DOM node object representing that element.  (In release 3 and earlier, this was named `elementID` and accepted only string values.)  This is the element which will have its contents replaced.  If the ID is not found in your document, SPARE throws an immediate exception.

> **`contentURL`** (required): the web address of the HTML content to be used for that replacement.  This can be a relative URL for content on the same site as the current page.  (Cross-domain URLs are commonly blocked by browser security anyway.)  Was named `pageURL` in release 3 and earlier.

> **`contentElementID`**:  the DOM ID of the element within the downloaded page which will be the source of the replacement content.  If you don’t provide any value (or pass a falsy value such as `""`) then it puts the entire content returned by the URL into your target element.  That technique is most appropriate if the server is set up to return fragmentary pages, instead of complete ones with `<html>` tags.  If a complete page is received and this ID is not given, it will use the content of the `<body>` tag.  (Extracting the body may be unreliable in Internet Explorer 9.  This problem does not affect any other IE version.)  Was named `newElementID` in release 3 and earlier.

> **`postData`**: values to be sent to the URL as form arguments.  If null or undefined, it requests the page with a simple GET.  Version 1 of SPARE supported only form-urlencoded strings, not multipart posts, and if you pass a string as this parameter, it still must be encoded in that format.  But version 2 supports alternate types of post data, including `URLSearchParams` and `FormData` objects if the browser is new enough to support them.  `URLSearchParams` does the form urlencoding for you, and `FormData` translates into multipart/form-data format, which supports file uploads.  (Passing in `ReadableStream`, `BufferSource`, or `Blob` objects is untested.)  `URLSearchParams` is less widely available than `FormData`, but has polyfills available.

> **`callbackContextData`**: an arbitrary object which is passed as the first argument to whichever of `onSuccess` or `onFailure` is invoked, so you can give them some information about the context in which they were called.

> **`onSuccess`**: a function to be invoked after the new content is successfully loaded.  If you pass a string, it will be executed with `eval`.  If it’s a function object, it will be passed the `callbackContextData` value.  For compatibility, values that are neither strings nor functions are simply ignored.  If defaulted, no action is taken after updating the content of your target element.  Any exceptions that occur in invoking it will pass through uncaught.  You can set a default value globally by assigning the function to the global variable **`SPARE.onSuccess`**.  This will be used if this parameter is unset or has a falsey value.

> **`onFailure`**: similar, but invoked if there’s a failure in loading the new content.  Again, the first argument is your `callbackContextData` value.  The second argument passed to it is an HTTP result number, such as 404 for page not found, and the third is the text of the error message received.  Note that with many websites transitioning to HTTP 2, the traditional status messages such as "Not Found" for 404 are no longer included in the response; if none is received then SPARE will now substitute a generic text such as "HTTP status 404", whereas SPARE 3 would just provide the empty string.  If the onFailure argument is not given, the default failure behavior is to navigate the browser window to the `contentURL` value passed in.  That is an appropriate fallback if you’re just using AJAX to smooth transitions during normal navigation, but  will not be useful if the server is returning only fragmentary pages.  Again, exceptions are not caught.  You can set a default value globally by assigning the function to the global variable **`SPARE.onFailure`**.  *NOTE*: the error number may also be negative, if SPARE fails to use the content after a successful download:

>> -1 means that `contentElementID` was not found in the downloaded content.

>> -3 means an unexpected exception was caught while importing the content into the document.  (-2 could be returned by SPARE version 1, but is not used anymore.)

>> 0 is also possible, with empty or meaningless text.  Requests do sometimes fail with no explanation available.  This can happen, for instance, when making a forbidden cross-site request.

> **`transitionalContentID`** \[IGNORED\]: a legacy parameter that no longer does anything, and is present only to retain API compatibility with earlier releases.  The global variable **`SPARE.transitionalContentID`** is also ignored.

> **`timeout`**: a number.  If the new data doesn’t download within this many seconds, the operation fails.  The supported range is from 0 to 3600, and the default is to leave it up to the browser.  Fractional values such as 1.25 are supported.  You can set a different default globally by putting a number in the global variable **`SPARE.timeout`**.  If the time expires, `onFailure` will be called with error code 408 (Request Timeout), with the error message being "SPARE time limit exceeded" instead of "Request Timeout".  (All internal messages to `onFailure` start with the word "SPARE".)  *Note* that setting a large value does not prevent the browser or server from failing the operation sooner.

All of these parameters except the first two are optional.  Furthermore, if the final parameter you pass in is a number, it will be taken as `timeout` even if other parameters before it are absent.  This means the function can be called with any one of the following additional polymorphic signatures:

```
SPARE.replaceContent(target, contentURL, timeout)
SPARE.replaceContent(target, contentURL, contentElementID, timeout)
SPARE.replaceContent(target, contentURL, contentElementID, postData, timeout)
SPARE.replaceContent(target, contentURL, contentElementID, postData, callbackContextData, timeout)
SPARE.replaceContent(target, contentURL, contentElementID, postData, callbackContextData, onSuccess, timeout)
SPARE.replaceContent(target, contentURL, contentElementID, postData, callbackContextData, onSuccess, onFailure, timeout)
```

--------

### simulateNavigation method

The second method is **`SPARE.simulateNavigation`**.  This works like `replaceContent` but has the additional effect of adding a history item under the browser’s Back button, and changing the URL visible in the browser’s address box.  This method is intended for a fairly strict and narrow case: when you replace part of a page’s content but wish to behave as if the entire page was replaced.  This makes sense if you have many pages that fit a common template.  The result of loading the partial page from the given URL should look the same as navigating to that page (only smoother); otherwise, using this method may be inappropriate, and produce results that are confusing to the page visitor.

*Note* that due to browser security, navigation between different domains will not work with `simulateNavigation`; all pages must be within a single website.  *Also note* that `simulateNavigation` does not at this time try to support use cases in which multiple sections of the page are independently updated.  Correctly implementing Back-button behavior in such cases is an order of magnitude more complex than in cases where just a single region is updated to hold various content.  I might attempt it in a future release.

The parameters of `simulateNavigation` mostly have the same meanings that they do when used with `replaceContent`.  Only the final two parameters, `newTitle` and `pretendURL`, are unique to `simulateNavigation`.  The latter was new in release 3.  The full list of parameters is:

> **`target`** (required): the ID of the existing HTML element which will have its contents replaced, or the DOM node representing that element.  (Formerly `elementID`.)

> **`contentURL`** (required): the URL from which new content will be loaded.  (Formerly `pageURL`.)

> **`contentElementID`**: the ID of the piece of the new content which will be loaded, or omit to use all of it.  (Formerly `newElementID`.)

> **`callbackContextData`**: arbitrary value passed to `onSuccess` or `onFailure`.

> **`onSuccess`**: callback function or expression invoked after new content is loaded.

> **`onFailure`**: callback function or expression invoked if the new content cannot be loaded.

> **`timeout`**: the number of seconds to wait for the download of new content.

> **`newTitle`**: a string which, if not blank, changes the title shown by the browser on the window or tab containing this page.

> **`pretendURL`**: a string which, if not blank, is shown in the address bar, and saved in the Back-button history, instead of `contentURL`.

Note that this method does not have the `postData` or `transitionalContentID` parameters that `replaceContent` offers.  Those features are not supported by this method.  We could in theory add `postData`, but it would be risky: the range of data that could be supplied to it may be sharply restricted, and it may not be possible to accurately pretest whether a post parameter will avoid causing exceptions later.

Also, at this time, `simulateNavigation` does not support omitting arguments before `timeout` the way that `replaceContent` does.

One gotcha to be aware of with `simulateNavigation` is that pretend URLs pushed into the history affect what path is “current” for relative URLs.  Unless all pages and pretend URLs are in the same directory, it’s safer to always use root-relative or absolute URLs throughout any pages that use SPARE.  I recommend root-relative URLs, because browser security does not permit us to simulate navigation to any other domains.

**IMPORTANT:** In order to use `simulateNavigation` effectively, you *must* set up a `popstate` event handler.  Otherwise, when the user clicks the Back button, the page content will not change!  SPARE includes a suitable handler which you can use or extend.

This event handler can be set by adding a line of code, either `window.onpopstate = SPARE.onPopStateRestore;` or `window.addEventListener('popstate', SPARE.onPopStateRestore);`.  Or you can substitute your own handler function, but its primary operation should be to call the supplied one.  Releases 2 and 3 of SPARE included a simple example of a popstate handler which you could use instead of the built-in one, but these simple versions have been found to be insufficient for some use cases.

--------

### onPopStateRestore event handler


ChangeLog.md
# SPARE Change Log

### Release 4 - August 3, 2022

1. Refactored the code to reduce repetition and bring the architecture closer to what will work in the future Promise-based version.

2. As part of refactoring, renamed input parameters `elementID` as `target`, `pageURL` as `contentURL`, and `newElementID` as `contentElementID`.

3. Added a feature to `simulateNavigation` and `onPopStateRestore` so they can simulate the `DOMContentLoaded` event.  This is activated by setting the new global property `SPARE.simulateDCL` to true.  If false it instead fires a new event called `SPAREContentLoaded`.

3. Improved `onPopStateRestore` to better handle returns to initially loaded pages, so that it can just replace the updated element with original content instead of reloading the whole page.  Added a safety check to it for cases where the history has somehow gotten out of sync.  (This may not be necessary.)  Gave `onPopStateRestore` a return value of `true` when it replaces content.  But clarified that `onPopStateRestore` is not yet ready for handling cases where multiple targets are updated, and it can’t be expected to correctly restore such documents.  The prior flaws in `onPopStateRestore` are the whole reason why I did an additional release of the old API — I couldn't leave release 3 as the only option for legacy users.

4. Updated the `state` object saved in browser history to use new renamed properties consistent with parameter naming elsewhere: `targetID` instead of `oldId`, `contentURL` instead of `url`, `contentElementID` instead of `newId`, `newTitle` instead of `title`, and `pretendURL` instead of `showURL`.  However, the poorly chosen old names are also still present for compatibility, until we transition to the new Promise-based API.  Also, `newTitle` and `pretendURL` and `targetID` are now present unconditionally.

5. Because HTTP 2 no longer includes a text description with its response status, such as "Not Found" for 404, SPARE 4 now substitutes a generic text such as "HTTP status 404" if none was received.

6. Changed the license from Apache to modified BSD.  My change is mainly to clarify the right to minify the script.

### Release 3 - June 12, 2021

_**NOTE:** The 2021 release of SPARE 3 has a known bug_ -- `window.title` _instead of_ `document.title` _in the line of_ `simulateNavigation` _which calls_ `history.replaceState` -- _which will cause problems if you use a global variable named_ `title`.

1. Added `pretendURL` parameter to `simulateNavigation`, so that the URL shown in the address bar can differ from the one that content is loaded from.  In popstate handlers, the state object now has a `showURL` property which stores this value.

2. When a popstate handler is returned to the page from which `simulateNavigation` was first called, it now receives a different state object with properties `startURL` and `startTitle`.  (In release 2, state would be null in this case, and the page content would probably fail to restore.)  The default handler `onPopStateRestore` now reloads the page from `startURL`.

3. Dropped support for the `transitionalContentID` feature.  The parameter and property are still present for API compatibility, but are now ignored.  (The API of SPARE 5 will be incompatible.)

4. Dropped support for IE 8 and 9, and any browser version older than about 2011.  This further simplifies the implementation by removing fallback code paths.  There are no longer any browsers where `supportLevel` is nonzero but `canSimulateNavigation` is false.

5. It now accepts HTTP result codes of 201 and 203 as well as 200.

### Release 2 - October 16, 2019

1. Support for Internet Explorer 7 (which was very limited) dropped completely, and with it, the return value of 1 from `supportLevel`.

2. Unified the XHR code paths so that only text mode is used, which turned out to make SPARE a lot faster.  Removed the return value of 3 from `supportLevel`.

3. Deprecated, but did not remove, the `transitionalContentID` parameter of the `replaceContent` method.  The API remains compatible with SPARE 1.

4. Added new API entry points `simulateNavigation` and `canSimulateNavigation`, and available event handler `onPopStateRestore`.

5. Added support for `FormData` and `URLSearchParams` objects being passed as `postData`, where release 1 only supported strings.

6. We now support `timeout` values between 0.0 and 1.0; in the first release, 1.0 seconds was the minimum allowed.

7. Supported polymorphic use of `timeout` in any optional parameter position, as long as the final parameter is a number.

8. Importing of downloaded HTML into document redone with better performance and fewer code paths.  This avoids a problem in release 1 where IE 9, 10, and 11 would, when asked to load a full HTML document, use content from the head as well as the body.  But *note* that though this problem is cured in IE 10 and 11, I cannot guarantee it is always mitigated in IE 9, particularly if the document you download has bad syntax.  (A future release will drop support for IE 8 and 9.)

### Release 1 - March 24, 2015

1. Created API entry points `replaceContent` and `supportLevel`.  A value of 1 returned by `supportLevel` indicates that only some functions work.

2. `XMLHttpRequest` (XHR) is used with two code paths, one using its text mode and the other using its XML/HTML mode.  The latter path is indicated by a return value of 3 from `SupportLevel`.


spare05X.js
// This is version ...5? of SPARE (Static Page AJAX for Replacing Elements), a JavaScript object.

/* Three-clause BSD-like license with simplified disclaimer and minification allowance:

"SPARE" is Copyright (c) 2015-2022 Paul M. Kienitz
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    Redistributions of source code must retain the above copyright notice
    and this list of conditions, including the following disclaimer:
    This software is provided "as is" without warranty of any kind.

    Redistributions in binary form must reproduce the above copyright
    notice and this list of conditions in the documentation and/or other
    materials provided with the distribution.  If the binary is incorporated
    into a work by URI reference, then this license may be referenced from
    the binary (e.g. minified script) via URI, provided such reference is
    immediately legible.

    The name of Paul Kienitz may not be used to endorse or promote products
    derived from this software without specific prior written permission.
*/


// In this version we move on from legacy APIs.  We ignore XHR in favor of fetch, and return a Promise.
// Most browsers only started supporting fetch in 2015-16, so older ones will not work without polyfills.
// IE 11 is the only legacy browser where this was tested with polyfills.

// Minify with https://www.digitalocean.com/community/tools/minify  (activate Eval and Safari10 options)

// TODO: INVESTIGATE unhandledrejection event.  TEST error handling in general.
//       ...Add hook to edit downloaded text as string before putting it in the dom?

"use strict";              // since there are polyfills for Promise and fetch, let's keep the language ES3-compatible
var SPARE = function ()	   // IIFE returns the SPARE singleton object, or null if unsupported
{
    // private read-only capability flag
    var canUseAbortController = "AbortController" in window;

    // private properties -- for thread safety, these are read-only after initialization
    var initialURL;
    var initialTitle;

    // private functions
    var makeError = function (exception, url, errorNumber, statusText)
    {
        if (exception && typeof(exception) === "string")   // for validation errors etc
            exception = new Error(exception);
        if (errorNumber > 0 && !statusText)
            statusText = "HTTP status " + errorNumber;     // for http2 where statusText is absent
        if (!exception && errorNumber)
            exception = new Error(errorNumber + " " + statusText);
        else if (!exception)
            exception = new Error(statusText);
        exception.httpStatus = errorNumber;
        exception.httpMessage = statusText;
        exception.contentURL = url;                 // so catch() handler can easily fall back to navigating there
        exception.isSPARE = true;                   // for unhandledrejection handlers that might get nonspare errors
        return exception;
    };

    var validate = function(target, contentURL)    // returns target DOM element if it doesn't throw
    {
        if (!contentURL || typeof(contentURL) !== "string")
            throw makeError("SPARE - contentURL is required", contentURL);
        if (target instanceof HTMLElement)
            return target;
        if (!target)
            throw makeError("SPARE - target is required", contentURL);
        var victim = document.getElementById(target);
        if (!victim)
            throw makeError("SPARE could not find target element '" + target + "'", contentURL);
        return victim;
    };

    var normalizeTimeout = function(timeout, timeout2)
    {
        if (isNaN(timeout))
            timeout = timeout2;
        if (isNaN(timeout) || timeout <= 0 || timeout > 3600)
            return undefined;
        else
            return timeout;
    };

    var extractAndUse = function (responseText, contentElementID, victim)       // returns error message or "" for success
    {
        var sideDocument = document.implementation.createHTMLDocument("");
        var newContentDomParent = sideDocument.documentElement;
        newContentDomParent.innerHTML = responseText;
        if (!contentElementID)
        {
            // When given a fragment to parse, documentElement generally wraps it in a simulated body tag.
            var body = newContentDomParent.getElementsByTagName("body");
            newContentDomParent = body[0] || newContentDomParent;
        }
        else           // find the named element
        {
            newContentDomParent = sideDocument.getElementById(contentElementID);
            if (!newContentDomParent)
                return "SPARE could not find element '" + contentElementID + "' in downloaded content";
        }

        var placeholder = document.createElement(victim.tagName);
        victim.parentNode.replaceChild(placeholder, victim);		// do the loops while detached from the dom, for performance
        while (victim.lastChild)
            victim.removeChild(victim.lastChild);
        while (newContentDomParent.firstChild)
            victim.appendChild(newContentDomParent.firstChild);
        placeholder.parentNode.replaceChild(victim, placeholder);
        return "";
    };


    // private internal classes
    var EventFirer = function(useDCL)
    {
        var eventName = useDCL ? "DOMContentLoaded" : "SPAREContentLoaded";
        this.fire = function ()
        {
            document.dispatchEvent(new Event(eventName, { bubbles: true }));
        };
    }

    var HistoryAdder = function (targetID, contentURL, contentElementID, newTitle, pretendURL)
    {
        var state = { targetID:         targetID,
                      contentURL:       contentURL,
                      contentElementID: contentElementID || null,
                      newTitle:         newTitle || null,
                      pretendURL:       pretendURL || null,
                      startTitle:       initialTitle || null,
                      startURL:         initialURL || null
        };
        var hindstate = { targetID:         targetID,
                          startTitle:       initialTitle || null,
                          startURL:         initialURL || null
        };
        this.add = function ()
        {
            history.pushState(state, "", pretendURL || contentURL);
            if (newTitle)
                document.title = newTitle;
        };
        this.checkBehind = function ()
        {
            if (!history.state)
                history.replaceState(hindstate, "");
        };
    };

    var Retriever = function (contentURL, postData, timeout, contentElementID, victim)
    {
        // private members -- per-transaction state is kept here
        var aborted = false;
        var timer = null;
        var fetchAborter = canUseAbortController ? new AbortController() : null;

        // our one internally public method
        this.start = function ()
        {
            var params;
            if (typeof postData === "string" || (postData !== null && typeof postData === "object"))
            {
                params = { method:  "POST",
                           body:    postData };
                if (typeof postData === "string")    // other parameter types set the header automatically
                    params.headers = { "Content-type": "application/x-www-form-urlencoded" };
            }
            else
                params = { method: "GET" };
            if (fetchAborter)
                params.signal = fetchAborter.signal;
            var promise = fetch(contentURL, params).then(fetchComplete).catch(fetchError);
            if (timeout)
                timer = setTimeout(abortBecauseTimeout, timeout * 1000);
            return promise;
        };

        // private methods
        var abortBecauseTimeout = function ()
        {
            aborted = true;
            if (fetchAborter)
                fetchAborter.abort();       // should halt HTTP session; in aborterless browsers, it continues silently in background
            downloadFailed(408, "SPARE time limit exceeded");
        };

        var fetchComplete = function (response)
        {
            if (!aborted)
            {
                clearTimeout(timer);
                if (response.status == 200 || response.status == 201 || response.status == 203)
                    return downloadSucceeded(response);    // this is the only path that doesn't throw
                else
                    downloadFailed(response.status, response.statusText);
            }
        };

        var fetchError = function (reason)
        {
            if (!aborted)                    // abort calls downloadFailed immediately, so don't do anything here
            {
                clearTimeout(timer);
                if ("name" in reason && "message" in reason)
                    downloadFailed(-2, "SPARE (during fetch) caught exception " + reason.name + ": " + reason.message, reason);
                else                         // should not happen?
                    downloadFailed(-4, "SPARE failed with reason " + reason);
            }
        };

        var downloadSucceeded = function (response)
        {
            try
            {
                var extractor = function (responseText)
                {
                    var err = extractAndUse(responseText, contentElementID, victim);
                    if (err)
                        downloadFailed(-1, err);
                    return victim;
                };
                return response.text().then(extractor).catch(fetchError);
            }
            catch (e)    // unlikely
            {
                downloadFailed(-3, "SPARE caught exception " + e.name + ": " + e.message, e);
            }
        };

        var downloadFailed = function (errorNumber, errorText, exception)
        {
            throw makeError(exception, contentURL, errorNumber, errorText);
        };
    };      // class Retriever


    // load-time initialization -- validate that we have browser support, and save initial location
    var supported = "fetch" in window && "Response" in window && "Promise" in window && "catch" in Promise.resolve(0) &&
                    "history" in window && "pushState" in history &&                 // ^^^ some polyfills don't set prototype
                    "implementation" in document && "createHTMLDocument" in document.implementation;
    // minimum browser versions without polyfills are from 2015-16: Edge 14, Firefox 39, Chrome 42, Safari 10.1
    // ...AbortController support is added 2017-19: Edge 16, Firefox 57, Chrome 66, Safari 12.1

    if (!initialURL)
    {
        initialURL = location.href;
        initialTitle = document.title;
    }

    // our IIFE result: create the SPARE object accessed by the caller, or set it null if the browser is lacking
    var spare = !supported ? null :
    {
        // global defaulting values settable by the caller
        timeout: undefined,
        simulateDCL: false,
        // XXX add global then-handler here?  or one just for onPopStateRestore?

        // Our core method - see https://github.com/paulkienitz/SPARE/blob/master/README.md for how to use.
        // Note that if you have ES8, going "await SPARE.replaceContent(...)" is an alternative to using .then().
        replaceContent: function (target /*ID or DOM element*/, contentURL, contentElementID, timeout, postData)
        {
            return new Promise(function (resolve, reject)
            {
                var victim = validate(target, contentURL);     // throws (which Promise turns into rejection) if no victim
                // Unlike SPARE 2-4, we do not support parsing the timeout parameter early when the last value passed in is a number.
                timeout = normalizeTimeout(timeout, SPARE.timeout);
                var retriever = new Retriever(contentURL, postData, timeout, contentElementID, victim);
                resolve(retriever.start());
            });
        },

        // Like replaceContent but also sets history and title.  No postData support.
        // HANDLER IS REQUIRED for popstate!  No cross-domain contentURL values are allowed
        // due to browser security.  Root-relative URLs are recommended.
        simulateNavigation: function (target, contentURL, contentElementID, timeout, newTitle, pretendURL)
        {
            var historyAdder, eventFirer, retriever;
            return new Promise(function (resolve, reject)
            {
                var victim = validate(target, contentURL);     // throws (which Promise turns into rejection) if no victim
                timeout = normalizeTimeout(timeout, SPARE.timeout);
                eventFirer = new EventFirer(SPARE.simulateDCL);
                historyAdder = new HistoryAdder(victim.id, contentURL, contentElementID, newTitle, pretendURL);
                historyAdder.checkBehind();

                retriever = new Retriever(contentURL, null, timeout, contentElementID, victim);
                resolve(retriever.start());
            }).then(historyAdder.add)
              .then(eventFirer.fire);
        },

        // This is a default handler for the popstate event, which can
        // be used with simulateNavigation if nothing fancier is needed, or
        // called by an extended handler to provide the core functionality.
        onPopStateRestore: function (event)
        {
            var ourPromise = undefined;
            if ("state" in event && event.state && "targetID" in event.state && "startURL" in event.state)
            {
                var eventFirer = new EventFirer(SPARE.simulateDCL);
                var victim = document.getElementById(event.state.targetID);
                // XXX Is this check unnecessary??
                if (!victim || location.href != (event.state.pretendURL || event.state.startURL))   // shouldn't happen
                {
                    console.log("=== SPARE had to reload initial page because assumed URL does not match current location, or target is missing." +
                                "\nPretend URL:  " + event.state.pretendURL + "\nInitial URL:  " + event.state.startURL +
                                "\n*Actual URL:  " + location.href + "\n- Target ID:  " + event.state.targetID);
                    location.replace(event.state.startURL);
                    // XXX what do we return here?
                }
                else if ("contentURL" in event.state)      // we are recreating a simulated non-original page state
                {
                    ourPromise = new Promise(function (resolve, reject)
                    {
                        var retriever = new Retriever(event.state.contentURL, null, SPARE.timeout, event.state.contentElementID, victim);
                        resolve(retriever.start());
                    });
                    document.title = event.state.title;
                }
                else                                       // we are returning to a page state as originally loaded
                {
                    ourPromise = new Promise(function (resolve, reject)
                    {
                        var retriever = new Retriever(event.state.startURL, null, SPARE.timeout, event.state.targetID, victim);
                        resolve(retriever.start());
                    });
                    document.title = event.state.startTitle;
                }
            }
            // XXX define a global function you can assign to be chained here when ourPromise is set??
            return ourPromise;
            // return value is ignored when this is used directly as the event handler
        }
    };      // the object literal that will be assigned to the global SPARE singleton
    return spare;
}();
